src/drivers/github.js

```diff
@@ -288,6 +288,112 @@ class Github {
     });
   }
 
+  async pipelineJobs(opts = {}) {
+    const { ids } = opts;
+    const { owner, repo } = ownerRepo({ uri: this.repo });
+    const { actions } = octokit(this.token, this.repo);
+
+    const jobs = await Promise.all(
+      ids.map(async (id) => {
+        const { data } = await actions.getJobForWorkflowRun({
+          owner,
+          repo,
+          job_id: id
+        });
+
+        return data;
+      })
+    );
+
+    return jobs.map((job) => {
+      const { id, started_at: date, run_id: runId, status } = job;
+      return { id, date, runId, status };
+    });
+  }
+
+  async job(opts = {}) {
+    const { time, status = 'queued' } = opts;
+    const { owner, repo } = ownerRepo({ uri: this.repo });
+    const { actions } = octokit(this.token, this.repo);
+
+    const {
+      data: { workflow_runs: workflowRuns }
+    } = await actions.listWorkflowRunsForRepo({
+      owner,
+      repo,
+      status
+    });
+
+    let runJobs = await Promise.all(
+      workflowRuns.map(async (run) => {
+        const {
+          data: { jobs }
+        } = await actions.listJobsForWorkflowRun({
+          owner,
+          repo,
+          run_id: run.id,
+          status
+        });
+
+        return jobs;
+      })
+    );
+
+    runJobs = [].concat.apply([], runJobs).map((job) => {
+      const { id, started_at: date, run_id: runId } = job;
+      return { id, date, runId };
+    });
+
+    const job = runJobs.reduce((prev, curr) => {
+      const diffTime = (job) => Math.abs(new Date(job.date).getTime() - time);
+      return diffTime(curr) < diffTime(prev) ? curr : prev;
+    });
+
+    return job;
+  }
+
+  async pipelineRestart(opts = {}) {
+    const { jobId: job_id } = opts;
+    const { owner, repo } = ownerRepo({ uri: this.repo });
+    const { actions } = octokit(this.token, this.repo);
+
+    const {
+      data: { run_id }
+    } = await actions.getJobForWorkflowRun({
+      owner,
+      repo,
+      job_id
+    });
+
+    try {
+      await actions.cancelWorkflowRun({
+        owner,
+        repo,
+        run_id
+      });
+    } catch (err) {
+      // HANDLES: Cannot cancel a workflow run that is completed.
+    }
+
+    const {
+      data: { status }
+    } = await actions.getWorkflowRun({
+      owner,
+      repo,
+      run_id
+    });
+
+    if (status !== 'queued') {
+      try {
+        await actions.reRunWorkflow({
+          owner,
+          repo,
+          run_id
+        });
+      } catch (err) {}
```

Comment:
What if it fails? Can we just ignore a specific error?

Reply:
If its fails there is nothing we can do. The only plausible scenario is connection issues.

Reply:
Yes, connection issues and backend errors would be the most plausible error causes there. Should we silently suppress them? If so, why?

Reply:
If thats the case that should be covered by the request done to retrieve the run_id. Additionally it adds the try because trying t restart and already restarting workflow throws an exception.

So:
 - If any error due to bad GH API state should be covered retrieving the run_id
 - We can add a log to show if the workflow failed to be retried if it already happened, very unlikely.

Reply:
> _Additionally it adds the try because trying to restart an already restarting workflow throws an exception._

And we can just ignore that exception and log it, without muting any other errors that may arise.

```suggestion
      } catch (err) {
        if (err.message === 'workflow is already restarting') // or whatever
          console.log(err.message) // or just ignore it
        else
          throw err
      }
```

> _If any error due to bad GH API state should be covered retrieving the run_id_

Unless, of course, there is a failure that only affects this API call. 🙃

Reply:
OK, but I consider this defensive programming to be honest. We are covering every case except them having broken that endpoint that will not change the result adding more code.

Reply:
When reproducing an issue requires ~5 minutes of server time and 30+ minutes of human time, it's usually nice to catch only those errors that we're ready to handle. We've talked a few times in past reviews about [deffensive programming](https://en.wikipedia.org/wiki/Defensive_programming), sometimes within reason and sometimes with opposite concepts in mind.

> _Defensive programming is a form of defensive design intended to ensure the continuing function of a piece of software under unforeseen circumstances._

> _Overly defensive programming, however, may safeguard against errors that will never be encountered, thus incurring run-time and maintenance costs._

> _There is also a risk that code traps prevent too many exceptions, potentially resulting in unnoticed, incorrect results._

I won't push too hard on this, because users won't hit that catch block for other reasons under normal circumstances, but, given the nature of CML, I would tend to fail as early as possible on those cases that we can't handle.

Perhaps I have acquired too many (cough) "bad" habits after using Python's selective `except` and following the recommendations.

Reply:
Moved to https://github.com/iterative/cml/issues/596#issuecomment-868719256.